### DISCUSSION TITLE
Tool change from tool table (no touch off)

### DISCUSSION THREAD

**Discussion by andrewmarles** [terjeio/grblHAL#323] [Q&A]
With quick change tooling I would like to set up the tool table in GRBLHAL and then manually (or with ATC someday) change tools and have it apply the tool offset from the table.  For example, set tool 1 as my probe/edge finder and use that to zero the work, and then have the tool change function apply offsets from there.  However, looking at the semi-auto tool change modes, these are all wanting either a manual or automatic touch off and this is what the quick change tooling (TTS for example) allows you to skip since the tool offsets are pre-defined.

Anybody have the tool table working in GRBLHAL, or is this something we need to add?

**Comment by langwadt** [comment]
is it not much simpler to do in the sender ?

**Reply by terjeio** [reply]
I don't think so - there are many senders out there. IMO better to keep the tool table and offset handling in the controller as all senders should be able to support that "out of the box" (unless they trap the associated gcodes).

When the tool table is [enabled](https://github.com/grblHAL/core/discussions/67#discussioncomment-1279183) in grblHAL [ioSender](https://github.com/terjeio/ioSender) supports sender side tool table management, the table itself and the associated gcodes are taken care of by the controller:

![image](https://user-images.githubusercontent.com/20260062/132083141-2e137a4e-c625-4f54-9e20-e8b4312d5ccf.png)

A bonus of having it controller side is that offline senders and/or SD-card streaming can make use of the tool table.
